### PR TITLE
Issue/4680 reader related traintracks

### LINKED ISSUE #4680: Reader: implement related posts train tracks in full post refresh
From @gibrown [here](https://github.com/wordpress-mobile/WordPress-Android/pull/4672#issuecomment-255473084): 

"Do the related posts here implement traintracks events and railcars per #4462? I couldn't find them in the code, but maybe they are elsewhere. Also, based on recent Calypso experience it would be good to have the traintracks render events only get fired when the related posts get shown to the user (so when the user has scrolled that far)."

### PR DESCRIPTION
**Pull request by nbradbury**
Resolves #4680 - updates tracking of related posts now that the "full post refresh" has been completed. There are two parts to this:

* Separately track when the user clicks a global or local related post (global being a related post from wp.com, local being a post from the same site as the current post)
* Track railcar render events for both global and local related posts (this tracking should only occur *after* the user has scrolled the related posts into view)

cc: @gibrown

### CODE REVIEW

**Comment by gibrown** [comment]
not that I know the Android codebase, but LGTM. Thanks!

File: WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostDetailFragment.java [hunk 1]
```diff
@@ -520,40 +539,84 @@ public void onEventMainThread(ReaderEvents.RelatedPostsUpdated event) {
...
+    private void showRelatedPostDetail(long blogId, long postId, boolean isGlobal) {
+        AnalyticsTracker.Stat stat = isGlobal
+                ? AnalyticsTracker.Stat.READER_GLOBAL_RELATED_POST_CLICKED
+                : AnalyticsTracker.Stat.READER_LOCAL_RELATED_POST_CLICKED;
+        AnalyticsUtils.trackWithReaderPostDetails(stat, blogId, postId);
 
         if (mIsRelatedPost) {
             mPostHistory.push(new ReaderBlogIdPostId(mPost.blogId, mPost.postId));
-            replacePost(blogId, postId, true);
+            replacePost(blogId, postId, false);
```
Comment (oguzkocer) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostDetailFragment.java]: I am not sure if I understand the reason of this change and the following. I looked into the commit but it's titled `Resolved conflicts`, so do you mind explaining the changes here just to make sure I don't miss anything. Thanks!

File: WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostDetailFragment.java [hunk 2]
```diff
@@ -520,40 +539,84 @@ public void onEventMainThread(ReaderEvents.RelatedPostsUpdated event) {
      * show the passed list of related posts - can be either global (related posts from
      * across wp.com) or local (related posts from the same site as the current post)
      */
-    private void showRelatedPosts(@NonNull ReaderRelatedPostList relatedPosts, boolean isGlobal) {
+    private void showRelatedPosts(@NonNull ReaderRelatedPostList relatedPosts, final boolean isGlobal) {
         // different container views for global/local related posts
         int id = isGlobal ? R.id.related_posts_view_global : R.id.related_posts_view_local;
```
Comment (nbradbury) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostDetailFragment.java]: Fixed in 6d6a55e

File: WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostDetailFragment.java [hunk 3]
```diff
@@ -520,40 +539,84 @@ public void onEventMainThread(ReaderEvents.RelatedPostsUpdated event) {
...
         }
+
+        trackRelatedPostsIfShowing();
+    }
+
+    /*
+     * track that related posts have loaded and are scrolled into view if we haven't
+     * already tracked it
+     */
+    private void trackRelatedPostsIfShowing() {
```
Comment (nbradbury) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostDetailFragment.java]: Argh, looks like I lost [this line](https://github.com/wordpress-mobile/WordPress-Android/blob/34d584ad1c8ec156e3b602c9375d04f04d3c9faa/WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostDetailFragment.java#L1065) when I merged `develop` into this branch. Re-added in 3b4303f

File: WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostDetailFragment.java [hunk 4]
```diff
@@ -520,40 +539,84 @@ public void onEventMainThread(ReaderEvents.RelatedPostsUpdated event) {
...
+    /*
+     * returns True if the passed view is visible and has been scrolled into view - assumes
+     * that the view is a child of mScrollView
+     */
+    private boolean isVisibleAndScrolledIntoView(View view) {
+        if (view != null && view.getVisibility() == View.VISIBLE) {
+            Rect scrollBounds = new Rect();
+            mScrollView.getHitRect(scrollBounds);
+            return view.getLocalVisibleRect(scrollBounds);
+        } else {
```
Comment (nbradbury) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostDetailFragment.java]: Fixed in 519d22a

File: WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostDetailFragment.java (hunk 1 above)
Comment (nbradbury) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostDetailFragment.java]: Glad you noticed that - it should be `true`. Fixed in 430c4e6

**Comment by nbradbury** [comment]
@oguzkocer Thanks for catching some really silly mistakes on my part. Not my best work, that's for sure. Ready for round two!

File: WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostDetailFragment.java [hunk 5]
```diff
@@ -520,40 +539,82 @@ public void onEventMainThread(ReaderEvents.RelatedPostsUpdated event) {
...
-                    mInterceptedUri);
+            ReaderActivityLauncher.showReaderPostDetail(
+                    getActivity(),
+                    false,
+                    blogId,
+                    postId,
+                    null,
+                    0,
+                    true,
+                    null);
```
Comment (oguzkocer) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostDetailFragment.java]: Should the last `null` be `mInterceptedUri` or is that a conscious change?

**Comment by oguzkocer** [comment]
@nbradbury I've added one more question, but aside from that I am ready to merge!

File: WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostDetailFragment.java (hunk 5 above)
Comment (nbradbury) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostDetailFragment.java]: That should be `null`. `mInterceptedUri` refers to the post the user is already viewing, not the related post they just tapped.

**Comment by oguzkocer** [comment]
Oops, forgot to give the :shipit: Looks great! Thanks for the changes @nbradbury 👍

**Pull request merged**